superset/alfred/utils.py

```diff
@@ -0,0 +1,350 @@
+                log["new event id"] = f"LS-<{event_type}_1>"
+
+    # Create Retention and Stage it
+    logger.info(f"Creating Retention...")
+    retention = RetentionInstance(
+        alfred,
+        f"Fission Retention - {datetime.now()}",
+        data_set,
+        metadata_system,
+        server_side_mapping_name="hogwarts.harmonized.eml_metadata",
+    )
+    ready = retention.stage()
+
+    # logger.info the result
+    logger.info(f"Retention Ready: {ready}.")
+    logger.info(f"Retention Messages: {retention.retention.message_summary}.")
+
+    # Return the URL
+    retention_url = (
+        InstanceUtil.build_url(alfred_instance)
+        .replace(":9488", "")
+        .replace("/rest", f"/ui/{retention.retention.uri}")
+    )
+
+    return retention_url
+
+
+def sanitize_results(data):
+    result = dict()
+    if isinstance(data, list):
+        json_array = []
+        for v in data:
+            json_array.append(sanitize_results(v))
+        return json_array
+    elif isinstance(data, NamedRowTuple):
+        for key, value in zip(data._names, data):
+            result[key] = sanitize_results(value)
+    else:
+        return data
+    return result
+
+def retain_eml_to_alfred(ids, alfred_env, access_token, dates=None):
+    try:
+        client = VaultClient()
+
+        ids_string = ""
+        datetimes = []
+        if ids:
+            logger.info(f'Found eml_ids: "{ids_string}" in arguments')
+            ids_string = "','".join(ids)
+        else:
+            raise Exception("No field 'email_ids' found in request")
+        if dates:
+            logger.info(f'Found dates: "{ids_string}" in arguments')
+            for date in dates:
+                try:
+                    datetimes.append(datetime.strptime(date, "%m-%d-%Y"))
+                except ValueError as ve:
+                    logger.info(ve)
+                    continue
+        if alfred_env:
+            logger.info(f'Found Alfred-env: "{alfred_env}" in arguments')
+        else:
+            raise Exception("No field 'alfred_env' found in request")
+
+        logger.info("Generating access token for trino...")
+        trino_app = client.get_oauth_client("trino")
+
+        trino_acces_token, trino_ctx = client.on_behalf_of(
+            f"{trino_app.audience}/{trino_app.scope}",
+            access_token,
+            token_client_name="superset",
+        )
+
+        trino_host = "trino-stg.hogwarts.pb.azure.chimera.cyber.gc.ca"
+        logger.info(f"Establishing connection to trino at {trino_host}...")
+        conn = connect(
+            auth=JWTAuthentication(trino_acces_token),
+            http_scheme="https",
+            host=trino_host,
+            port=443,
+            catalog="hogwarts_pb",
+        )
+
+        sql = f"""
+    SELECT id, urls, "from", to, message_id, reply_to, stream_id, subject, eml_path, bcc, cc, ministerial_authorization, classification, time
+    FROM  hogwarts_pb.harmonized.eml_metadata
```

Comment:
All of this relies on that schema and table? Just checking.

Reply:
yep!

Reply:
And field names! 🔥

Reply:
Is there a way to externalize this that we should be considering?